ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/TransactionalProducerFactory.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.rest.server;
+
+import io.confluent.ksql.rest.server.computation.CommandRunner;
+import java.util.Map;
+import java.util.Objects;
+import org.apache.kafka.clients.producer.ProducerConfig;
+
+public class TransactionalProducerFactory {
```

Comment:
commandStore would need to keep track of the commandRunner in order to pass it to each TransactionalProducer in `createTransactionalProducer`. I'm not sure how exactly that would work out because the commandStore is created first in RestApp and then the commandRunner.

Reply:
Why can't we reuse the existing mechanism in `CommandStore` to do this? (see the implementation of `completeSatisfiedSequenceNumberFutures`). Then the producer factory has no dependency on `CommandRunner`.

Reply:
The CommandStore version introduces a race condition. The futures for the sequenceNumbers are completed before the commands are actually executed in commandRunner. There's a bit of lag between consuming and actually executing the command and updating the metastore. Also, the CommandRunner could get stuck executing consumed commands, but ensureConsumedPast() in CommandStore wouldn't block the TransactionalProducer.waitForConsumer() call and then the transaction would proceed on an out of sync metastore.

Reply:
After thinking about it and playing around with the existing functionality of `CommandStore.ensureComsumedPast()`, it fits with what I was trying to achieve by duplicating the code. I removed the code from CommandRunner and the factory class. I'm not sold on dropping `enqueueCommand()` from commandQueue though since the `CommandStore.commandStatusMap` is being updated in `enqueueCommand()`.

Reply:
> I'm not sold on dropping enqueueCommand() from commandQueue though since the CommandStore.commandStatusMap is being updated in enqueueCommand()

Is the concern that you won't be able to get to the map from the queue? In this way of doing things the implementation of the producer interface could be _inside_ CommandStore, so that shouldn't be a problem. So something like:

```
interface TransactionalProducer {
    send()
}

interface CommandQueue {
    getProducer()
}

class CommandStore {
    Producer getProducer() {
        return new TopicProducer();
    }
    class TopicProducer implements TransactionalProducer {
        send() { ... }
    }
}
```

This way, the underlying queue is still encapsulated away.

Reply:
Mind if I leave this for a future PR? I think this PR has gone through enough refactors already.